File: 2020/23/Main.hs
import qualified Part1                         as P1
import qualified Part2                         as P2
import           System.Environment
import           Text.Printf

output :: String -> IO ()
output path = do
  input <- head . lines <$> readFile path
  printf "File: %s\n" path
  printf "  Part 1: %s\n" . P1.part1 $ input
  (r1, r2, p) <- P2.solve input
  printf "  Part 2: %d * %d = %d\n" r1 r2 p

main :: IO ()
main = getArgs >>= mapM_ output

File: 2020/23/Rust/src/main.rs
mod part_1;
mod part_2;

use part_1 as P1;
use part_2 as P2;

use std::convert::TryFrom;
use std::env;
use std::fs;

fn parse_input(input: &str) -> Vec<i32> {
    let u32_to_i32 = |x: &u32| i32::try_from(*x).unwrap();
    input
        .lines()
        .take(1)
        .collect::<String>()
        .chars()
        .map(|x| u32_to_i32(&x.to_digit(10).unwrap()))
        .collect()
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    println!("File: {}", path);
    println!("  Part 1: {}", P1::solve(&parsed));
    let (r_1, r_2, p) = P2::solve(&parsed);
    println!("  Part 2: {} * {} = {}", r_1, r_2, p);
}

fn main() {
    env::args().skip(1).for_each(|x| output(&x));
}
